docs/configuration.md
# Configuration

Netvisor integration credentials must be supplied in order for the upstream API integration to work. The backend application reads configuration from environment variables according to the table below.

Variables with default values can be omitted from environment.

For local development, create `.env` file in repository root and populate it with your demo credentials.

## Environment Variables

<!-- prettier-ignore -->
Key|Default|Description
-|-|-
`NODE_ENV`||Node environment spec. Value **MUST BE** `production` whenever you are using non-demo credentials for Netvisor.
`NETVISOR_API_URL`|https://isvapi.netvisor.fi|URL where Netvisor API requests should be made to. Default value points to NV's demo environment.
`NETVISOR_CUSTOMER_ID`||See Netvisor auth docs for `X-Netvisor-Authentication-CustomerId` header.
`NETVISOR_ORGANIZATION_ID`||See Netvisor auth docs for `X-Netvisor-Organisation-ID` header.
`NETVISOR_CUSTOMER_KEY`||See Netvisor auth docs for _"Integraatiokäyttäjän yksilöivä avain"_.
`NETVISOR_ORGANIZATION_KEY`||See Netvisor auth docs for _"Integraatiokumppanin yksilöivä avain"_.
`NETVISOR_PARTNER_ID`||See Netvisor auth docs for `X-Netvisor-Authentication-PartnerId` header.
`NETVISOR_CACHE_TTL`|60|Time-to-live (seconds) for caching certain Netvisor API endpoint results.
`NETVISOR_ENTRY_RATIONUMBER`|100|Ratio number ("collectorratio") to use when making new workday entries.
`PORT`|3001|Server listens to this port.
`EMPLOYEE_NUMBER_HEADER_KEY`|X-SHIB-employeeId|Name of the header that defines authenticated user's Netvisor employee ID (case insensitive). **The user must not be able to set this header.**
`LOG_LEVEL`<sup>1</sup>|log|Filter log output by suppressing log messages of higher level than the given value. Log levels in ascending order are `error`, `warn`, `log` and `debug`.
`ENABLE_JSON_LOGS`<sup>1</sup>|`false`|Output stringified JSON instead of human-readable logs by setting this to `true`.
`ENABLE_AUDIT_LOGS`<sup>1</sup>|`false`|When set to `true`, logs of level `audit` are printed, containing users' personal information. These logs should be stored safely or not enabled at all in production environments. Audit log output is always JSON and does not respect the `LOG_LEVEL` setting.
`DATABASE_USERNAME`||Postgres username.
`DATABASE_PASSWORD`||Postgres password.
`DATABASE_NAME`||Postgres database name.
`DATABASE_HOST`||Postgres hostname.
`DATABASE_PORT`||Postgres port.
`DATABASE_SSL_MODE`|`false`|When set to `true`, TypeORM is configured to use SSL mode `no-verify` when connecting to Postgres. Otherwise, SSL is disabled. See [`node-postgres` docs](https://github.com/brianc/node-postgres/tree/master/packages/pg-connection-string#tcp-connections) for more.
`ATLASSIAN_CLIENT_ID`|| Client ID of your Atlassian OAuth 2.0 App. See [Jira OAuth2.0 apps](https://developer.atlassian.com/cloud/jira/platform/oauth-2-3lo-apps/).
`ATLASSIAN_CLIENT_SECRET`|| Client Secret of your Atlassian OAuth 2.0 App.
`ATLASSIAN_AUTHORIZATION_URL`|https://auth.atlassian.com/authorize| Atlassian URL where user is directed to grant keijo access to use resources.
`ATLASSIAN_TOKEN_URL`|https://auth.atlassian.com/oauth/token| Atlassian URL where keijo exchanges authorization code for access and refresh tokens.
`ATLASSIAN_CLOUD_ID`|| Cloud ID of Atlassian tenant where Keijo requests data from.
`CALLBACK_URL`|/jira/callback| Keijo URL where user is redirected from Jira after access is granted.
`CALLBACK_REDIRECT_URL`|/| Keijo URL where user is redirected from callback after callback is handled.
`SESSION_SECRET`|| Session secret.
`CORS_ORIGIN`|| Development Cors origin URL.
`TRUST_PROXY_IPS`|false| Trusted proxy ips to allow secure cookies to be sent over proxies.
`DEV_TOOLS`|`false`|Set to `true` to enable dev tools for development and testing. This cannot be set to `true` if `NODE_ENV` is `production` unless `CI` is also set to `true`.
`CI`|`false`|Indicates that server is running in an CI environment.

<sup>1</sup>) See [Logging docs](./logging.md) for more information.


server/src/config/config.ts
const {
  NODE_ENV,
  PORT,
  EMPLOYEE_NUMBER_HEADER_KEY,
  LOG_LEVEL,
  ENABLE_JSON_LOGS,
  ENABLE_AUDIT_LOGS,
  NETVISOR_CUSTOMER_ID,
  NETVISOR_ORGANIZATION_ID,
  NETVISOR_CUSTOMER_KEY,
  NETVISOR_ORGANIZATION_KEY,
  NETVISOR_PARTNER_ID,
  NETVISOR_CACHE_TTL,
  NETVISOR_ENTRY_RATIONUMBER,
  NETVISOR_API_URL,
  DATABASE_USERNAME,
  DATABASE_PASSWORD,
  DATABASE_NAME,
  DATABASE_HOST,
  DATABASE_PORT,
  DATABASE_SSL_MODE,
  ATLASSIAN_TOKEN_URL,
  ATLASSIAN_CLIENT_ID,
  ATLASSIAN_CLIENT_SECRET,
  ATLASSIAN_AUTHORIZATION_URL,
  ATLASSIAN_CLOUD_ID,
  CALLBACK_URL,
  CALLBACK_REDIRECT_URL,
  SESSION_SECRET,
  TRUST_PROXY_IPS,
} = process.env;

const config = {
  nodeEnv: NODE_ENV,
  inDev: NODE_ENV?.toLowerCase() === "development",
  port: Number(PORT) || 3001,
  employeeNumberHeaderKey: EMPLOYEE_NUMBER_HEADER_KEY || "X-SHIB-employeeId",
  logLevel: LOG_LEVEL?.toLowerCase() || "log",
  enableJsonLogs: ENABLE_JSON_LOGS?.toLowerCase() === "true",
  enableAuditLogs: ENABLE_AUDIT_LOGS?.toLowerCase() === "true",
  netvisor: {
    host: NETVISOR_API_URL || "https://isvapi.netvisor.fi",
    sender: "keijo",
    partnerId: NETVISOR_PARTNER_ID,
    lang: "fi",
    customerId: NETVISOR_CUSTOMER_ID,
    organizationId: NETVISOR_ORGANIZATION_ID,
    customerKey: NETVISOR_CUSTOMER_KEY,
    organizationKey: NETVISOR_ORGANIZATION_KEY,
    cacheTtl: Number(NETVISOR_CACHE_TTL) || 60,
    ratioNumber: Number(NETVISOR_ENTRY_RATIONUMBER) || 100,
  },
  database: {
    username: DATABASE_USERNAME,
    password: DATABASE_PASSWORD,
    name: DATABASE_NAME,
    host: DATABASE_HOST,
    port: Number(DATABASE_PORT),
    ssl: DATABASE_SSL_MODE === "true",
  },
  jira: {
    tokenUrl: ATLASSIAN_TOKEN_URL || "https://auth.atlassian.com/oauth/token",
    clientId: ATLASSIAN_CLIENT_ID,
    clientSecret: ATLASSIAN_CLIENT_SECRET,
    authorizationUrl: ATLASSIAN_AUTHORIZATION_URL || "https://auth.atlassian.com/authorize",
    callbackUrl: CALLBACK_URL || "/jira/callback",
    callbackRedirectUrl: CALLBACK_REDIRECT_URL || "/",
    scopes: "read:jira-work offline_access",
    cloudId: ATLASSIAN_CLOUD_ID,
  },
  session: {
    secret: SESSION_SECRET,
    name: "sessionId",
  },
  trustProxyIps: TRUST_PROXY_IPS || false,
};

export default config;
